log: Add option to re-open a log file

chrissie-c opened the pull request:
With an optional new filename

chrissie-c commented:
This changes the API of the custom loggers (to take a struct timespec rather than a time_t) but we're already changing the soname of the library anyway for v2.0.0

wferi commented:
> This changes the API of the custom loggers (to take a struct timespec rather than a time_t)

How does time representation influence log reopening? Thanks.

chrissie-c commented:
> > This changes the API of the custom loggers (to take a struct timespec rather than a time_t)
> 
> How does time representation influence log reopening? Thanks.

It doesn't - this is a different PR I'm still working on, sorry. Blame it on Mondays!

chrissie-c commented on lib/log_file.c:
```diff
@@ -69,11 +69,14 @@ static void
+		fclose(oldfile);
 	}
-	t->instance = fopen(t->filename, "a+");
+	newfile = fopen(t->filename, "a+");
```
Good points all. I've pushed a new commit to fix them

wferi commented on lib/log_file.c:
```diff
@@ -132,13 +148,14 @@ int32_t
+	if (rc) {
 		return -errno;
 	} else {
 		return 0;
```
Why not move this part into `_do_file_reload()` so that you can simply `return _do_file_reload(t)` here? Gets rid of a conditional and moves reading of `errno` much closer to the function setting it, which adds to clarity IMO.

chrissie-c replied:
We can't do that because _do_file_reload() is called from the logger object and that doesn't return an error (bad design, I know, not guilty). That's why I had to split it into two functions in the first place!

wferi replied:
You mean `_file_reload()`, don't you? You could still define it to `errno = -_do_file_reload(t, NULL)` I guess (using the second argument for passing in the new filename as discussed in my other comment). Then get rid of it altogether with the next API change.

wferi commented on lib/log_file.c:
```diff
@@ -132,13 +148,14 @@ int32_t
+	int rc;
 
 	if (filename) {
 		(void)strlcpy(target->filename, filename, PATH_MAX);
```
Isn't changing the filename field premature here? If `open()` fails, we keep the old `FILE` structure after all. I overlooked this added complication; looks like one must pass the new file name to `_do_file_reload()` in a separate argument...

wferi replied:
I believe you. It might be useful to close this avenue of future confusion anyway.

chrissie-c commented:
I've changed the reopen to only close the old file if the new one succeeds, and moved the fclose to be sure that errno is retained on return.

Return code from _file_reload() I'm going to leave as-is for the moment. Changing the logger api is a bigger job and should be kept al together IMO.

wferi commented on lib/log_file.c:
```diff
@@ -147,13 +153,9 @@ qb_log_file_close(int32_t t)
-		(void)strlcpy(target->filename, filename, PATH_MAX);
-	}
-	rc = _do_file_reload(t);
+	rc = _do_file_reload(filename, t);
```
Why not `return -errno; /* from fopen() */` instead of `return -1;` in `_do_file_reload()` so that you can then rename it to `int32_t qb_log_file_reopen()` and happily be one function down?

wferi commented on lib/log_file.c:
```diff
@@ -60,20 +60,45 @@ _file_close(int32_t t)
-		fclose(t->instance);
+	/* Only close oldfile if newfile succeeds */
+	if (oldfile && newfile) {
+		fclose(oldfile);
```
I'd write this as `if (oldfile) { fclose(oldfile); }` inside the `if (newfile) { }` block below for clarity. YMMV.

wferi commented:
Wouldn't dropping support for changing the log file name during reopen help a little? That's somewhat unrelated anyway and remains possible by stopping and starting logging the old way.

jfriesse commented:
@wferi Not too much. Current logging thread is simply not protected from changing configuration at all (so qb_log_ctl2) and errors are visible even when something as simple as reload corosync.conf is performed (it basically works just by accident).

chrissie-c commented:
To clarify what's happening here. I force pushed a new patch yesterday that adds locking around all of the logging reconfiguration. Today's small amendment is to make 'in_logger' atomic as it's accessed across threads and helgrind was complaining (it still complains but only because it only understands pthreads).

jfriesse commented:
Last update seems to solve all problems with `corosync-cfgtool -R` and logging , so ACK by me.

Looks like some of @wferi's comments are still unresolved (Why not return -errno; ...) so probably not yet ready to go.

chrissie-c commented:
Thanks. I'd rather leave it as it is TBH, then the two 'API' functions are equal and the fn that does the work is called by both.

wferi commented:
Your call, certainly. I probably wouldn't have brought it up if `_do_file_reload()` wasn't introduced in this very pull request. If you prefer to have this symmetry between `_file_reload()` and `qb_log_file_reopen()`, then be it.

chrissie-c commented:
Thanks.
